Mathlib/Geometry/Manifold/VectorField.lean

```diff
@@ -675,4 +675,368 @@ end ContMDiff
+      (mpullbackWithin 𝓘(𝕜, E) I (extChartAt I x₀).symm W (range I))
+      ((extChartAt I x₀).symm ⁻¹' s ∩ range I)) x₀
+
+variable (I I') in
+/-- The Lie bracket of two vector fields in a manifold. -/
+def mlieBracket (V W : Π (x : M), TangentSpace I x) (x₀ : M) : TangentSpace I x₀ :=
+  mlieBracketWithin I V W univ x₀
+
+lemma mlieBracketWithin_def  :
+    mlieBracketWithin I V W s = fun x₀ ↦
+    mpullback I 𝓘(𝕜, E) (extChartAt I x₀)
+    (lieBracketWithin 𝕜
+      (mpullbackWithin 𝓘(𝕜, E) I (extChartAt I x₀).symm V (range I))
+      (mpullbackWithin 𝓘(𝕜, E) I (extChartAt I x₀).symm W (range I))
+      ((extChartAt I x₀).symm ⁻¹' s ∩ range I)) x₀ := rfl
+
+lemma mlieBracketWithin_apply :
+    mlieBracketWithin I V W s x₀ =
+    (mfderiv I 𝓘(𝕜, E) (extChartAt I x₀) x₀).inverse
+    ((lieBracketWithin 𝕜
+      (mpullbackWithin 𝓘(𝕜, E) I (extChartAt I x₀).symm V (range I))
+      (mpullbackWithin 𝓘(𝕜, E) I (extChartAt I x₀).symm W (range I))
+      ((extChartAt I x₀).symm ⁻¹' s ∩ range I)) ((extChartAt I x₀ x₀))) := rfl
+
+lemma mlieBracketWithin_eq_lieBracketWithin {V W : Π (x : E), TangentSpace 𝓘(𝕜, E) x} {s : Set E} :
+    mlieBracketWithin 𝓘(𝕜, E) V W s  = lieBracketWithin 𝕜 V W s := by
+  ext x
+  simp [mlieBracketWithin_apply]
+
+/-********************************************************************************
+Copy of the `lieBracket` API in manifolds
+-/
+
+@[simp] lemma mlieBracketWithin_univ :
+    mlieBracketWithin I V W univ = mlieBracket I V W := rfl
+
+lemma mlieBracketWithin_eq_zero_of_eq_zero (hV : V x = 0) (hW : W x = 0) :
+    mlieBracketWithin I V W s x = 0 := by
+  simp only [mlieBracketWithin, mpullback_apply, comp_apply]
+  rw [lieBracketWithin_eq_zero_of_eq_zero]
+  · simp
+  · simp only [mpullbackWithin_apply]
+    have : (extChartAt I x).symm ((extChartAt I x) x) = x := by simp
+    rw [this, hV]
+    simp
+  · simp only [mpullbackWithin_apply]
+    have : (extChartAt I x).symm ((extChartAt I x) x) = x := by simp
+    rw [this, hW]
+    simp
+
+lemma mlieBracketWithin_swap_apply :
+    mlieBracketWithin I V W s x = - mlieBracketWithin I W V s x := by
+  rw [mlieBracketWithin, lieBracketWithin_swap, mpullback_neg]
+  rfl
+
+lemma mlieBracketWithin_swap :
+    mlieBracketWithin I V W s = - mlieBracketWithin I W V s := by
+  ext x
+  exact mlieBracketWithin_swap_apply
+
+lemma mlieBracket_swap_apply : mlieBracket I V W x = - mlieBracket I W V x :=
+  mlieBracketWithin_swap_apply
+
+lemma mlieBracket_swap : mlieBracket I V W = - mlieBracket I W V :=
+  mlieBracketWithin_swap
+
+@[simp] lemma mlieBracketWithin_self : mlieBracketWithin I V V = 0 := by
+  ext x; simp [mlieBracketWithin, mpullback]
+
+@[simp] lemma mlieBracket_self : mlieBracket I V V = 0 := by
+  ext x; simp_rw [mlieBracket, mlieBracketWithin_self, Pi.zero_apply]
+
+/-- Variant of `mlieBracketWithin_congr_set` where one requires the sets to coincide only in
+the complement of a point. -/
+theorem mlieBracketWithin_congr_set' (y : M) (h : s =ᶠ[𝓝[{y}ᶜ] x] t) :
```

Comment:
Taking a step back: can you briefly say why this generality is useful?

Reply:
This is a good question, but also a good answer.